File: lua/vim-calc/vim-calc.lua
local luaxp = require("vim-calc.deps.luaxp")

local M = {}

local function calcExpr(expr)
  -- ---------------------------------------------------
  -- This function takes an expression and calculates the
  -- resulting value.
  -- ---------------------------------------------------
  -- compile parses the expression
  local parsedExp, cerr = luaxp.compile(expr)
  -- handle failures in parsing
  if parsedExp == nil then
    if cerr == nil then
      print("Parsing failed, cannot get error message.")
      return
    else
      print("Parsing failed: " .. cerr.message)
      return
    end
  end
  
  -- run calculates the expression
  local resultValue, rerr = luaxp.run(parsedExp)
  -- handle errors in calculation
  if resultValue == nil then
    if rerr == nil then
      print("Evaluation failed, cannot get error message.")
      return
    else
      print("Evaluation failed: " .. rerr.message)
      return
    end
  else
    return resultValue
  end
end

local function getSelectedRange()
  -- ---------------------------------------------------
  -- This function finds the current selected range in
  -- the buffer. It returns the start and end position 
  -- in order, even if the selection was made by 
  -- selecting backwards.
  -- ---------------------------------------------------
  -- get the selection positions
  local pos1 = vim.fn.getpos("v")
  local pos2 = vim.fn.getpos(".")
  -- order positions based on their relative values
  -- handles selecting forward and backwards
  if pos1[2] >= pos2[1] and pos1[3] >= pos2[3] then
    return pos2, pos1
  else
    return pos1, pos2
  end
end

local function getSelectedText()
  -- ---------------------------------------------------
  -- This function gets the currently selected text in
  -- visual mode. Can handle multiple lines.
  -- ---------------------------------------------------
  local select_start, select_stop = getSelectedRange()
  -- extract the start and end row and column
  local start_line, start_column = select_start[2], select_start[3]
  local stop_line, stop_column = select_stop[2], select_stop[3]
  -- get the lines of the buffer that are selected
  local lines = vim.api.nvim_buf_get_lines(0, start_line - 1, stop_line, false)
  
  -- handle no selection, single line, and multi line
  local text = ""
  if #lines == 0 then
    text = ""
  elseif #lines == 1 then
    text = string.sub(lines[1], start_column, stop_column)
  else
    lines[1] = string.sub(lines[1], start_column)
    lines[#lines] = string.sub(lines[#lines], 1, stop_column)
    text = table.concat(lines, "\n")
  end
  return text, select_start, select_stop
end

function M.calcSelectedText()
  -- ---------------------------------------------------
  -- This function retrieves the selected text from the
  -- buffer, uses luaxp to parse and calculate the 
  -- resulting value, and replaces the selected text
  -- with the resulting value.
  -- ---------------------------------------------------
  local text, select_start, select_stop = getSelectedText()
  local s_buf, s_row, s_col, _ = unpack(select_start)
  local _, e_row, e_col, _ = unpack(select_stop)
  local result = calcExpr(text)
  vim.api.nvim_buf_set_text(
    s_buf,
    s_row - 1,
    s_col - 1,
    e_row - 1,
    e_col,
    {}
  )
  vim.api.nvim_buf_set_text(
    s_buf,
    s_row - 1,
    s_col - 1,
    s_row - 1,
    s_col - 1,
    { string.format("%.4f", result) }
  )
  vim.api.nvim_feedkeys(
    vim.api.nvim_replace_termcodes(
      '<esc>',
      true,
      false,
      true),
    'x', false
  )
end

return M

File: python3/vimcalc/vimcalc.py
import vim
from datetime import datetime
from asteval import Interpreter

def get_timestamp(dt=None):
    if dt:
        return datetime.isoformat(dt)
    else:
        return datetime.isoformat(datetime.now())

def log(level, message):
    ts = get_timestamp()
    output = f"[{level.upper()}] [vimcalc] [{ts}] : {message}"
    print(output)

def strings_to_ints(strings):
    return [int(i) for i in strings]

def get_selected_range():
    pos1 = strings_to_ints(vim.eval("getpos('v')"))
    pos2 = strings_to_ints(vim.eval("getpos('.')"))
    log("debug", f"pos1 = {pos1}; pos1 = {pos2}")
    if pos1[1] >= pos2[1] and pos1[2] >= pos2[2]:
        return pos2, pos1
    else:
        return pos1, pos2

def get_text_from_range(start, stop):
    start_line, start_column = start[1], start[2]
    stop_line, stop_column = stop[1], stop[2]
    lines = vim.eval(f"nvim_buf_get_lines(0, {start_line - 1}, {stop_line}, v:false)")
    if len(lines) == 0:
        return ""
    elif len(lines) == 1:
        return lines[0][start_column - 1:stop_column]
    else:
        lines[0] = lines[0][start_column - 1:]
        lines[-1] = lines[-1][:stop_column]
        return "\n".join(lines)

def preprocess_input(text):
    return text.lstrip(" ").rstrip(" ") 

def evaluate_expression(expr):
    aeval = Interpreter()
    try:
        value = float(aeval(expr))
    except Exception:
        value = expr
    return value

def replace_buffer_range(start, stop, replacement):
    s_buf, s_row, s_col, _ = start
    _, e_row, e_col, _ = stop
    set_text_args = [s_buf, s_row - 1, s_col - 1, e_row - 1, e_col, []]
    set_text_arg_string = ", ".join(str(i) for i in set_text_args)
    vim.eval(f"nvim_buf_set_text({set_text_arg_string})")
    set_text_args = [s_buf, s_row - 1, s_col - 1, s_row - 1, s_col - 1, [replacement]]
    set_text_arg_string = ", ".join(str(i) for i in set_text_args)
    vim.eval(f"nvim_buf_set_text({set_text_arg_string})")
    feedkeys = vim.eval("nvim_replace_termcodes('<esc>', v:true, v:false, v:true)")
    vim.eval(rf"nvim_feedkeys('{feedkeys}', 'x', v:false)")

def print_with_python():
    start, stop = get_selected_range()
    text = get_text_from_range(start, stop)
    log("debug", f"Selected text: {text}")
    expr = preprocess_input(text)
    log("debug", f"Preprocessed text: {expr}")
    result = evaluate_expression(expr)
    log("debug", f"Result: {result}")
    if isinstance(result, float):
        output = f"{result:.4}"
    else:
        output = result
    replace_buffer_range(start, stop, output)
